### ISSUE TITLE
[Question] How to debug "Timeout to process message"

### ISSUE THREAD

**Issue by burnpanck** [thingsboard/thingsboard#9625]
**Component**

 Rule Engine

**Description**
Our ThingsBoard instance seems to randomly stop working every few weeks. The instance in question is tiny with less than 100 devices in total, most of them are idle for most of the time. Yet, at some point, 100 % of the telemetry messages start to get lost, even at minimal load (< 10 messages per minute). CPU and Memory load on the VM remain low at all times, and the web-UI remains responsive. The telemetry messages are successfully accepted, but seem to fail somewhere in ThingsBoard's queue. The following is an excerpt of the symptom from the ThingsBoard logfile:
```
2023-11-13 15:09:34,319 [sql-log-1-thread-1] INFO  o.t.s.dao.sql.TbSqlBlockingQueue - Queue-1 [Attributes] queueSize [0] totalAdded [1] totalSaved [1] totalFailed [0]
2023-11-13 15:09:37,882 [tb-rule-engine-consumer-35-thread-6 | QK(Main,TB_RULE_ENGINE,system)-7] INFO  o.t.s.s.q.DefaultTbRuleEngineConsumerService - Timeout to process [2] messages
2023-11-13 15:09:37,883 [tb-rule-engine-consumer-35-thread-6 | QK(Main,TB_RULE_ENGINE,system)-7] INFO  o.t.s.s.q.DefaultTbRuleEngineConsumerService - [bef89a60-f8a1-11ed-82e1-5d1088348e82] Timeout to process message: TbMsg(queueName=Main, id=501769d2-cbb1-4d94-a216-4d7883e66588, ts=1699888145073, type=POST_TELEMETRY_REQUEST, originator=fbb9a4d0-7a1a-11ee-884d-71513c3a2dd7, customerId=e2422c70-7a1a-11ee-884d-71513c3a2dd7, metaData=TbMsgMetaData(data={...redacted...}, ruleChainId=null, ruleNodeId=null, ctx=org.thingsboard.server.common.msg.TbMsgProcessingCtx@19eb521f, callback=org.thingsboard.server.common.msg.queue.TbMsgCallback$1@1590e8a4), Last Rule Node: null
2023-11-13 15:09:39,883 [tb-rule-engine-consumer-35-thread-6 | QK(Main,TB_RULE_ENGINE,system)-7] INFO  o.t.s.s.q.DefaultTbRuleEngineConsumerService - Timeout to process [2] messages
2023-11-13 15:09:39,883 [tb-rule-engine-consumer-35-thread-6 | QK(Main,TB_RULE_ENGINE,system)-7] INFO  o.t.s.s.q.DefaultTbRuleEngineConsumerService - [bef89a60-f8a1-11ed-82e1-5d1088348e82] Timeout to process message: TbMsg(queueName=Main, id=a8ba7626-259c-4d5e-8c94-7fc9ce8ca217, ts=1699888165077, type=POST_TELEMETRY_REQUEST, originator=fbb9a4d0-7a1a-11ee-884d-71513c3a2dd7, customerId=e2422c70-7a1a-11ee-884d-71513c3a2dd7, metaData=TbMsgMetaData(data={...redacted...}, ruleChainId=null, ruleNodeId=null, ctx=org.thingsboard.server.common.msg.TbMsgProcessingCtx@435fa1a5, callback=org.thingsboard.server.common.msg.queue.TbMsgCallback$1@1590e8a4), Last Rule Node: null
2023-11-13 15:09:43,287 [sql-log-1-thread-1] INFO  o.t.s.dao.sql.TbSqlBlockingQueue - Queue-2 [TS] queueSize [0] totalAdded [8] totalSaved [8] totalFailed [0]
```

After updating the instance to 3.6.1 and thereby restarting, the symptoms are gone and the instance works again normally. A previous incident also resolved after a restart of the instance. However, we would still prefer if we could prevent that from happening in the first place.

**Environment**

 * OS:  Ubuntu 20.04 LTS
 * ThingsBoard: 3.5.1 CE
 * Queue: In-Memory
 * DB: PostgreSQL

**Comment by albtrentadue** [comment]
Hi - I have exactly the same issue.
We also have very LOW load, less than 20 devices, 1 telemetry every 30s, no load on the server now swap.
Here is the log:
`2023-11-19 23:10:27,913 [tb-rule-engine-consumer-63-thread-25 | QK(Main,TB_RULE_ENGINE,system)-2] INFO  o.t.s.s.q.DefaultTbRuleEngineConsumerService - [30343d00-3aa8-11ed-9110-252a46c628c3] Timeout to process message: TbMsg(queueName=Main, id=adcb6a4f-17b4-49e1-8773-c8d8178f71b3, ts=1700431824798, type=POST_TELEMETRY_REQUEST, originator=31407930-5716-11ee-a487-4f9464f05293, customerId=8a6c1920-63b3-11ee-a487-4f9464f05293, metaData=TbMsgMetaData(data={deviceType=Filtro osmosi inversa, deviceName=SN01135, ts=1700431824000}), dataType=JSON, data={ ...removed... }, ruleChainId=8227daf0-40e2-11ed-aaa5-1d7ab9bee6e2, ruleNodeId=8260ec50-40e2-11ed-aaa5-1d7ab9bee6e2, ctx=org.thingsboard.server.common.msg.TbMsgProcessingCtx@26b380ce, callback=org.thingsboard.server.common.msg.queue.TbMsgCallback$1@1594576f), Last Rule Node: [RuleChain: Filtri Osmosi Processamento Telemetrie|RuleNode: Check presenza errore warning(8260ec50-40e2-11ed-aaa5-1d7ab9bee6e2)]
`
Only difference is that "Last Rule Node:" here is not "null"

We would like to know
1. How to find out reason for timeout
2. What to change in queues to avoid this timeout

These reports reduce perceived reliability by users.

OS: Ubuntu 20.04 LTS
ThingsBoard: 3.4.4 PE
Queue: In-Memory
DB: PostgreSQL

**Comment by trikimiki** [comment]
Hi!

`Timeout to process message(s)` usually mean that, as the text suggests, the processed message hit the timeout during processing in the specific Rule Chain.
If there is a specific "place" specified, like `Last Rule Node: [RuleChain: rc_name|RuleNode: rn_name(rn_id)]` - timeout happened in the specified node; if not, like `Last Rule Node: null` - message was timed out upon "transitioning" between two Rule Nodes.

As for debugging those timeouts, there is no unified flow, because potential root causes are too broad. The most generic causes/fixes are listed next, but the best way to debug is a general check of debug events and ThingsBoard app logs (what happens before/after timeout).  

There are multiple ways of fixing this, the most common I can suggest:
- check server/cluster performance metrics (performance "bottleneck" -> more time required to process data -> timeouts)
- check Rule Engine queue settings, tweak timeout options if needed
- if you do not want to tweak timeout queue-wide, you can use the "checkpoint" node to switch queues - this also resets the timeout.

Note that, depending on your Rule Engine Queue parameters, timed-out messages may be re-processed. The more severe errors are "Permanent timeouts" (message hit the re-process attempts count) and "Failures" (script errors usually). The best way for checking those is API Usage -> Rule Engine -> "graph button" dashboard available for Tenant Admins.

@burnpanck your case indicates either server performance issues or misconfigured timeout parameter for the queue. If first is not an issue, check your Rule Engine Queue settings under Sys Admin.

@albtrentadue your case indicates that processing in the specific Rule Node is slow (maybe, in the specific Rule Chain "branch"). This is true if the timeout errors always point to the same Rule Node(s). The best way to fix this is to either tweak Rule Engine Queue overall, or use "checkpoint" as I have suggested before.

**Comment by burnpanck** [comment]
> your case indicates either server performance issues or misconfigured timeout parameter for the queue. If first is not an issue, check your Rule Engine Queue settings under Sys Admin.

I did check those, but they look fine to me. Also, we never changed those settings from the default. Specifically, the `main` queue is configured as follows:
 - Strategy type: Burst
 - Processing type: "Retry failed and timeout", 3 retries within 3 s each (I may have adjusted the processing type from the default after the second time I observed the behaviour)
 - Poll interval 25, 10 partitions
 - Send message poll for each consumer: yes, processing within 2000 ms

Also, the system works initially, and then stops working, so it feels more like a type of internal resource exhaustion.

> ... but the best way to debug is a general check of debug events and ThingsBoard app logs (what happens before/after timeout).

As I said, the instance is *extremely* idle in the observed situation. In my case, it appears that there is nothing at all happening before/after, at least as far as I can tell from the application log. However, documentation on debug log is quite limited. I found instructions [how to enable "certain" logs](https://thingsboard.io/docs/user-guide/troubleshooting/#enable-certain-logs), but nothing on what "certain" logs I should enable. Is there any logger/class[hierarchy] specifically you would suggest to enable in this case?
